ops/opset13/rnn.go

```diff
@@ -0,0 +1,292 @@
+	activationBeta  []float32
+	activations     []string
+	direction       RNNDirection
+	hiddenSize      int
+}
+
+// newRNN creates a new rnn operator.
+func newRNN() ops.Operator {
+	return &RNN{
+		activations: []string{"tanh"},
+		direction:   Forward,
+	}
+}
+
+// Init initializes the rnn operator.
+func (r *RNN) Init(attributes []*onnx.AttributeProto) error {
+	for _, attr := range attributes {
+		switch attr.GetName() {
+		case "activation_alpha":
+			r.activationAlpha = attr.GetFloats()
+		case "activation_beta":
+			r.activationBeta = attr.GetFloats()
+		case "activations":
+			activations := []string{}
+			for _, activation := range attr.GetStrings() {
+				activations = append(activations, string(activation))
+			}
+
+			r.activations = activations
+		case "clip":
+			return ops.ErrUnsupportedAttribute(attr.GetName(), r)
+		case "direction":
+			r.direction = RNNDirection(attr.GetS())
+			if r.direction != Forward {
+				return ops.ErrUnsupportedAttribute(attr.GetName(), r)
+			}
+		case "hidden_size":
+			r.hiddenSize = int(attr.GetI())
+		default:
+			return ops.ErrInvalidAttribute(attr.GetName(), r)
+		}
+	}
+
+	return nil
+}
+
+// Apply applies the rnn operator.
+func (r *RNN) Apply(inputs []tensor.Tensor) ([]tensor.Tensor, error) {
+	if inputs[4] != nil {
+		return nil, ops.ErrUnsupportedInput("sequence lens", r)
+	}
+
+	X := inputs[0]
+
+	Wi, err := r.getWeights(inputs[1])
+	if err != nil {
+		return nil, err
+	}
+
+	Ri, err := r.getWeights(inputs[2])
+	if err != nil {
+		return nil, err
+	}
+
+	B := inputs[3]
+	if B == nil {
+		B = r.getDefaultB()
+	}
+
+	Wbi, Rbi, err := r.getBiases(B)
+	if err != nil {
+		return nil, err
+	}
+
+	seqLength := X.Shape()[0]
+	batchSize := X.Shape()[1]
+
+	Ht := inputs[5]
+	if Ht == nil {
+		Ht = r.getInitialH(batchSize)
+	}
+
+	// Reshape the hidden tensor without the bidirectional dimension, as
+	// we do not support bidirectional RNN yet. This is the first dimension.
+	if err = Ht.Reshape(Ht.Shape().Clone()[1:]...); err != nil {
+		return nil, err
+	}
+
+	activation := RNNActivations[r.activations[0]]
+	if activation == nil {
+		return nil, ops.ErrUnsupportedAttribute("activations", r)
+	}
+
+	outputs := []tensor.Tensor{}
+
+	// Loop over all timesteps of the input, applying the RNN calculation to every
+	// timesteps while updating the hidden tensor.
+	for t := 0; t < seqLength; t++ {
+		Xt, err := X.Slice(ops.NewSlicer(t, t+1), nil, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		Ht, err = r.layerCalculation(Xt, Ht, Wi, Ri, Wbi, Rbi, RNNActivations[r.activations[0]])
+		if err != nil {
+			return nil, err
+		}
+
+		outputs = append(outputs, Ht)
+	}
+
+	Y := outputs[0]
+	if len(outputs) > 1 {
+		Y, err = tensor.Concat(0, Y, outputs[1:]...)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	Yh, ok := Ht.Clone().(tensor.Tensor)
+	if !ok {
+		return nil, ops.ErrTypeAssert("tensor.Tensor", Ht.Clone())
+	}
+
+	// Reshape the outputs so it adds the num_directions as specified by onnx.
```

Comment:
You mention `num_directions` in the comment but it is not clear to me where you use this?